testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java

```diff
@@ -75,29 +83,84 @@
     private SqlParser sqlParser;
     private final AutoCloseableCloser afterClassCloser = AutoCloseableCloser.create();
     private io.trino.sql.query.QueryAssertions queryAssertions;
+    private Map<Handler, Level> oldLevels;
+    private MemoryHandler memoryHandler;
 
     @BeforeClass
     public void init()
             throws Exception
     {
+        enableLogBuffering();
         queryRunner = afterClassCloser.register(createQueryRunner());
         h2QueryRunner = afterClassCloser.register(new H2QueryRunner());
         sqlParser = new SqlParser();
         queryAssertions = new io.trino.sql.query.QueryAssertions(queryRunner);
     }
 
+    private void enableLogBuffering()
+    {
+        if (System.getenv("CONTINUOUS_INTEGRATION") == null) {
+            return;
+        }
+        // change the log level of the console handler
+        Logger rootLogger = Logger.getLogger("");
+        Handler[] handlers = rootLogger.getHandlers();
+        oldLevels = new HashMap<>();
+        Handler consoleHandler = null;
+        for (Handler handler : handlers) {
+            if (!(handler instanceof ConsoleHandler)) {
+                continue;
+            }
+            oldLevels.put(handler, handler.getLevel());
+            handler.setLevel(Level.OFF);
+            consoleHandler = handler;
+        }
+
+        // add a capturing handler
+        if (consoleHandler != null) {
+            memoryHandler = new MemoryHandler(consoleHandler, 1_000_000, Level.OFF);
+            rootLogger.addHandler(memoryHandler);
+        }
+    }
+
     protected abstract QueryRunner createQueryRunner()
             throws Exception;
 
     @AfterClass(alwaysRun = true)
-    public final void close()
+    public final void close(ITestContext context)
             throws Exception
     {
         afterClassCloser.close();
         queryRunner = null;
         h2QueryRunner = null;
         sqlParser = null;
         queryAssertions = null;
+        flushBufferedLogs(context);
+    }
+
+    private void flushBufferedLogs(ITestContext context)
+    {
+        if (System.getenv("CONTINUOUS_INTEGRATION") == null) {
+            return;
+        }
+        if (oldLevels.size() == 0) {
+            return;
+        }
+        // restore the log level of the console handler
+        for (Handler handler : Logger.getLogger("").getHandlers()) {
+            if (!(handler instanceof ConsoleHandler) || !oldLevels.containsKey(handler)) {
+                continue;
+            }
+            handler.setLevel(oldLevels.get(handler));
+        }
+        if (memoryHandler == null) {
+            return;
+        }
+        if (context.getFailedTests().size() != 0) {
```

Comment:
I'll check. I was not aware if we prefer JUnit 5 or TestNG.

Reply:
There's a [TestWatcher](https://junit.org/junit5/docs/5.4.0/api/org/junit/jupiter/api/extension/TestWatcher.html). Is there any existing JUnit 5 test that has a Query Runner or logs something?

Reply:
We’re trying to move to JUnit 5: https://github.com/trinodb/trino/issues/9378

Reply:
`AbstractTestQueryFramework` fundamentally requires TestNG due to the annotations in the base class, so this would have to be updated when we migrate it.

Reply:
Indeed, but I wouldn’t want to add dependencies that will make it harder or introduce a roadblock to migrating.

Reply:
Well, maybe I can start the migration now? Would this be a good enough reason?